Move serialize_program and serialize_run_context into EngineContext

maffoo opened the pull request:
It's convenient to be able to use these functions outside of having an `EngineProgram` object or calling `Engine.create_program`, for example when testing direct use of the streaming API with the combined `CreateQuantumProgramAndJobRequest` call.

@wcourtney, @verult

wcourtney commented on cirq-google/cirq_google/engine/engine.py:
```diff
@@ -126,6 +126,28 @@ def copy(self) -> 'EngineContext':
+        program = serializer.serialize(program)
+        return _pack_any(program)
+
+    def serialize_run_context(self, sweeps: List[cirq.Sweep], repetitions: int) -> any_pb2.Any:
```
Since these are now public, can you add unit tests to cirq-google/cirq_google/engine/engine_test.py?

wcourtney commented on cirq-google/cirq_google/engine/engine.py:
```diff
@@ -126,6 +126,28 @@ def copy(self) -> 'EngineContext':
     def _value_equality_values_(self):
         return self.proto_version, self.client
 
+    def serialize_program(
```
I find the name "program" confusing here. Since it's moving to public visibility, what do you think of making it something like `serialize_circuit` and let the weird naming in `create_program` be its own problem?

maffoo replied:
The term "program" is fairly well-established, for example it is used in all the [`cirq.Sampler`](https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/work/sampler.py#L34) methods, so I'm inclined to leave this, but I could be persuaded.

wcourtney commented on cirq-google/cirq_google/engine/engine.py:
```diff
@@ -126,6 +126,28 @@ def copy(self) -> 'EngineContext':
+        program = serializer.serialize(program)
+        return _pack_any(program)
+
+    def serialize_run_context(self, sweeps: List[cirq.Sweep], repetitions: int) -> any_pb2.Any:
```
Should this also accept a serializer for consistency with the `serialize_program` method above?

maffoo replied:
The serializer is only used for circuits, so not needed for serializing the run context.

maffoo commented:
I reworked this a bit when thinking about testing:
- moved the core logic of populating `RunContext` and `BatchRunContext` into new helper functions in `v2/sweeps.py` where we already have helpers for converting sweeps to protos. Added some tests for these new functions.
- moved the code for packing messages into `Any` protos into a helper function in a new `engine.util` module. This replaces a bunch of duplicate versions of this logic in tests.
- made the new helpers on `EngineContext` private. These really aren't meant to be "public" functions, just wanted them to be accessible in testing without having an engine program object.

wcourtney commented on cirq-google/cirq_google/api/v2/sweeps_test.py:
```diff
@@ -134,3 +134,37 @@ def test_sweep_with_flattened_sweep():
     assert v2.sweep_to_proto(param_sweep2) is not None
+
+
+def test_run_context_to_proto():
```
Can you also add a test to demonstrate that specifying the `out` parameter is appended to? Same for the batch case.

maffoo replied:
Done.

wcourtney commented on cirq-google/cirq_google/engine/util.py:
```diff
@@ -0,0 +1,25 @@
+from google.protobuf.message import Message
+
+
+def pack_any(message: Message) -> any_pb2.Any:
```
Shocking that there's not a standard utility for this, but it doesn't look like there is...

maffoo replied:
I know, right?

wcourtney reviewed (approved):
Nice rework! LGTM % additional test coverage.
